### PR TITLE
Add Serde get/set JSON functionality to Rust examples/docs

### PR DESCRIPTION
**Pull request by tpmccallum**
Content must go through a pre-merge checklist.

## Pre-Merge Content Checklist

This documentation has been checked to ensure that:

- [x] The `title, `template`, and `date` are all set
- [x] Does this PR have a new menu item (anywhere in `templates/*.hbs` files) that points to a document `.md` that is set to publish in the future? If so please only publish the `.md` and `.hbs` changes in real-time (otherwise there will be a menu item pointing to a `.md` file that does not exist)
- [x] File does not use CRLF, but uses plain LF (hint: use `cat -ve <filename> | grep '^M' | wc -l` and expect 0 as a result) 
- [x] Has passed [`bart check`](https://developer.fermyon.com/bartholomew/quickstart)
- [x] Has run `npm run build-index`
- [x] Has been manually tested by running in Spin/Bartholomew (hint: use `PREVIEW_MODE=1` and run `npm run styles` to update styling)
- [x] Headings are using Title Case
- [x] Code blocks have the programming language set to properly highlight syntax and the proper copy directive
- [x] Have tested with `npm run test` and resolved all errors
- [x] Relates to an existing (potentially outdated) blog article? If so please add URL in blog to point to this content.

### CODE REVIEW

**Comment by tpmccallum** [comment]
Closes https://github.com/fermyon/developer/issues/614

File: content/spin/kv-store-api-guide.md
Comment (tpmccallum) [inline content/spin/kv-store-api-guide.md:1]: The first commit just removes the warnings that were already present on a new build:

```
warning: unused import: `Error`
 --> src/lib.rs:5:17
  |
5 |     key_value::{Error, Store},
  |                 ^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused variable: `req`
 --> src/lib.rs:8:19
  |
8 | fn handle_request(req: Request) -> Result<Response> {
  |                   ^^^ help: if this is intentional, prefix it with an underscore: `_req`
  |
  = note: `#[warn(unused_variables)]` on by default
```

File: content/spin/rust-components.md [hunk 1]
```diff
@@ -367,6 +367,61 @@ messages on the `messages` Redis channel.
...
+    // Open the default key-value store
+    let store = Store::open_default()?;
+    // Define a User object using serde derive
+    #[derive(Serialize, Deserialize, Debug)]
+    struct User {
+        fingerprint: String,
+        location: String,
+    }
+    // Create an instance of a user object and populate the values
+    let user = User{fingerprint:"0x1234".to_owned(), location:"Brisbane".to_owned()};
```
Comment (rylev) [inline content/spin/rust-components.md:392]: nit: can we format these snippets with rustfmt?

File: content/spin/rust-components.md [hunk 2]
```diff
@@ -367,6 +367,61 @@ messages on the `messages` Redis channel.
...
+    // Store the user object using the "my_json" key
+    store.set_json(String::from("my_json"), &user)?;
```
Comment (rylev) [inline content/spin/rust-components.md:394]:
```suggestion
    store.set_json("my_json", &user)?;
```

File: content/spin/rust-components.md [hunk 3]
```diff
@@ -367,6 +367,61 @@ messages on the `messages` Redis channel.
...
+    // Retrieve the user object from the key-value store, using the "my_json" key
+    let retrieved_user: User = store.get_json("my_json")?;
+    // Respond to the request by revealing the user's fingerprint as the response body
+    Ok(http::Response::builder().status(200).body(Some(retrieved_user.fingerprint.into()))?)
+}
+```
+
+The above example requires some additional configuration. Specifically, enabling the Spin SDK's optional `json` feature (from v1.2.0 onwards) and adding Serde's optional `derive` feature in the `dependencies` section of the application's `Cargo.toml` file:
```
Comment (itowlson) [inline content/spin/rust-components.md:402]: The Rust templates don't reference `serde` by default so suggest rephrasing the second part, but I'm not sure if this is too wordy:

```suggestion
To use the `get_json` and `set_json` helpers, you must enable the Spin SDK's optional `json` feature in your `Cargo.toml` file.  To make your objects serialisable, you will also need a reference to `serde`; the example above also requires Serde's `derive` feature.  The relevant `Cargo.toml` entries look like this:
```

I wonder if it's worth adding a "details" expander along the lines of "Why are the helpers under an optional feature? // `get_json` and `set_json` depend on the `serde` and `serde_json` libraries. These are large libraries, so applications that depend on them are larger and take longer to build.  We didn't want users of the Rust SDK to incur these costs unless they actually wanted the feature."  Not ideal wording I fear but hopefully you get the idea!

**Review by itowlson — CHANGES_REQUESTED** [review]
Looks good - a few suggestions but mostly fiddly or just moving stuff around.

File: content/spin/rust-components.md (hunk 1 above)
Comment (tpmccallum) [reply content/spin/rust-components.md:392]: Sure think; sorry, I should have done that in the first place.
Formatting will be in the next commit.

File: content/spin/kv-store-api-guide.md [hunk 4]
```diff
@@ -68,6 +69,9 @@ fn handle_request(req: Request) -> Result<Response> {
 `open` **Operation**
 - The close operation is not surfaced; it is called automatically when the store is dropped.
 
+`set_json` and `get_json` **Operation**
+- Rust applications can [store and retrieve Rust data structures](spin/rust-components#storing-data-in-the-spin-key-value-store).
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/kv-store-api-guide.md:73]: I have gone with "serializable Rust types", as you have suggested because the Serde documentation refers to that same nomenclature when talking about Serde's out-of-the-box functionality. Which I guess is as far as this documentation goes; just top-level Serde usage.

File: content/spin/kv-store-tutorial.md [hunk 5]
```diff
@@ -176,9 +176,11 @@ command = "tinygo build -target=wasi -gc=leaking -no-debug -o main.wasm main.go"
 
 In this section, we use the Spin SDK to open and persist our application's data inside our default key/value store. This is a special store that every environment running Spin applications will make available for their application. 
 
+> Please note: Spin applications written in Rust can [store and retrieve Rust data structures](spin/rust-components#storing-data-in-the-spin-key-value-store) in the application's data store.
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/kv-store-tutorial.md:179]: Agreed, happy to let this one stay here for visibility - for now. And not be a precedent for littering.

File: content/spin/rust-components.md [hunk 6]
```diff
@@ -367,6 +367,61 @@ messages on the `messages` Redis channel.
...
+    key_value::{Store},
+};
+use serde::{Serialize, Deserialize};
+
+#[http_component]
+fn handle_request(_req: Request) -> Result<Response> {
+    // Open the default key-value store
+    let store = Store::open_default()?;
+    // Define a User object using serde derive
+    #[derive(Serialize, Deserialize, Debug)]
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/rust-components.md:386]: Removed `Debug`, thanks!

File: content/spin/rust-components.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/rust-components.md]: Thanks, I have hoisted that up and it looks better. Good call.

File: content/spin/rust-components.md [hunk 7]
```diff
@@ -367,6 +367,61 @@ messages on the `messages` Redis channel.
 
 Spin has a key-value store built in. For information about using it from Rust, see [the key-value store tutorial](kv-store).
 
+If you want to go beyond just inserting values in a certain spot, Spin's Rust SDK provides an optional feature for using [Serde](https://docs.rs/serde/latest/serde). The Rust code below shows how to store and retrieve serializable objects from the key-value store:
```
[the remark this answers is not in the document]
Comment (tpmccallum) [reply content/spin/rust-components.md]: Yes, another good call. This page is long and with this `H3` we get some clarity and get to see it in the ToC.
Updated in latest commit.

File: content/spin/rust-components.md [hunk 8]
```diff
@@ -367,6 +368,72 @@ messages on the `messages` Redis channel.
...
+```rust
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+use spin_sdk::{
+    http::{Request, Response},
+    http_component,
+    key_value::Store,
+};
+
+// Define a User object using serde derive
```
Comment (itowlson) [inline content/spin/rust-components.md:396]:
```suggestion
// Define a serializable User type
```

File: content/spin/rust-components.md [hunk 9]
```diff
@@ -367,6 +368,72 @@ messages on the `messages` Redis channel.
...
+        .body(Some(retrieved_user.fingerprint.into()))?)
+}
+```
+
+Once built and running (using `spin build` and `spin up`) you can test the above example in your browser (by visiting localhost:3000) or via curl, as shown below:
+
+<!-- @selectiveCpy -->
+
+```bash
+$ curl -i localhost:3000
```
Comment (itowlson) [inline content/spin/rust-components.md:429]: Suggestion: 

```suggestion
$ curl localhost:3000
```

and remove the status/header lines from the output - they don't really show anything.

**Review by itowlson — APPROVED** [review]
Looks good - thanks!  I left a couple of cosmetic suggestions but they are certainly not crucial.
